lib/knapsack_pro/formatters/time_tracker.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+module KnapsackPro
+  module Formatters
+    class TimeTracker
+      ::RSpec::Core::Formatters.register self,
+        :example_group_started,
+        :example_started,
+        :example_finished,
+        :example_group_finished,
+        :stop
+
+      # Called at the beginning of each subset,
+      # but only the first instance of this class is used,
+      # so don't rely on the initializer to reset values.
+      def initialize(_output)
+        @time_each = nil
+        @time_all = nil
+        @before_all = 0.0
+        @group = {}
+        @subset = {}
+        @queue = {}
+        @queue_started = now
+        @subset_started = now
+      end
+
+      def example_group_started(notification)
+        return unless top_level_group?(notification.group)
+        @time_all = now
+      end
+
+      def example_started(_notification)
+        @before_all = now - @time_all if @before_all == 0.0
+        @time_each = now
+      end
+
+      def example_finished(notification)
+        record_example(@group, notification.example, @time_each)
+        @time_all = now
+      end
+
+      def example_group_finished(notification)
+        return unless top_level_group?(notification.group)
+
+        add_hooks_time(@group, @before_all, now - @time_all)
+        @subset = merge(@subset, @group)
+        @before_all = 0.0
+        @group = {}
+      end
+
+      # Called at the end of each subset
+      def stop(_notification)
+        @queue = merge(@queue, @subset)
+        @subset = {}
+        @subset_started = now
+      end
+
+      def queue(scheduled_paths)
+        recorded_paths = @queue.values.map do |example|
+          example[:path].match(/\A(.*?)(?:\[([\d\s:,]+)\])?\z/).captures.first
+        end
+
+        missing = (scheduled_paths - recorded_paths).each_with_object({}) do |path, object|
+          object[path] = { path: path, time_execution: 0.0 }
+        end
+
+        merge(@queue, missing).values.map do |example|
+          example.transform_keys(&:to_s)
+        end
+      end
+
+      def subset
+        @subset.values.map do |example|
+          example.transform_keys(&:to_s)
+        end
+      end
+
+      def queue_time
```

Comment:
```suggestion
      def total_time
```

Maybe we could use a more general name. If the formatter is used in queue mode and regular mode then from the formatter perspective it does not have to know the queue mode name.